### ISSUE TITLE
Possible bug in the flow analysis of nullable reference types

### ISSUE THREAD

**Issue by johnknoop** [dotnet/csharplang#2882]
I've just enabled the new nullable reference types feature.

In this example, `myObj` is nullable. But shouldn't the compiler understand that myObj won't be null in this case, since I'm testing for it in the if clause?

![image](https://user-images.githubusercontent.com/3394270/66725159-cc007880-ee2e-11e9-8720-04b72fd709b0.png)

I'm using .NET Core 3.0 in VS 2019.

**Comment by orthoxerox** [comment]
It should, but the compiler cannot infer everything. In 8.0 it recognizes the most common null-check patterns only. Does the warning go away if you replace the test with `myObj != null && myObj.HasChanged`?

**Comment by YairHalberstadt** [comment]
You can report this on Roslyn, and they may choose to add support for this pattern.

**Comment by theunrepentantgeek** [comment]
I agree that this is a pattern that the language/compiler should likely support.

I can [eproduce the warning [using SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFDIMQDsCuAbbBDAI2wFMACY9Qk5KAZlNlIGFTkBvZUrh+qAFlIBZABRMA/KQDGASlZJupDvIXcAlgDNSwyWIB0ACTwBnJgAs86AObFgpMRPV5sR4rM4quSjx4AqAJwsjAFsAe0s/DQBPbV0ASWBpAG53DwBfFNJ05S4MugY4ADYGAX9A0PCo4QBlABcIq1JVW2kMr24shVz6AhCQ7FJDE3MrG0VSaxrEzNIAXlI6zGJk7J58gAZSeLGJqdTZ0gAiJoPl1KA=).

Changing the condition from `?? false` to `== true`  has the same semantics, and doesn't generate a warning.

**Comment by alrz** [comment]
Relates to discussion at https://github.com/dotnet/csharplang/issues/2764

I still don't think it makes sense to pinpoint to such "patterns" inside the compiler, this is a byproduct of how flow analysis works. Ideally, it should be extended in a general manner to understand always-true/false/null in a broader context. 

R# has done a good job to infer such states in the local scope *even* when the code is not annotated at all.

**Comment by johnknoop** [comment]
Another thing I would expect to work is when you initialize a member using `??`.

In the example below, I get this warning:

> Non-nullable property 'MyProp' is uninitialized. Consider declaring the property as nullable.

```csharp
public class MyClass
{
    public MyClass()
    {
        MyProp ??= Some.Value();
    }

    public string MyProp { get; private set; }
}
````

If I initialize it using `=` instead of `??=`, the warning goes away.

**Comment by YairHalberstadt** [comment]
@johnknoop

Again, I think you can report that on roslyn. Can I ask why you would do that though?

**Comment by johnknoop** [comment]
When the class is initially created, there are a few properties that aren't being set (the value isn't produced yet). Hence they have to be nullable.

Later on, these values are added to the persisted MongoDB document representing the class, in a way that only uses expressions and never instantiates the class.

When loaded, I don't wanna do null-testing in all sites of usage, so I added the "if null" assignments to the constructor, so the values of the properties are either a real value or a suitable default value.

**Comment by Joe4evr** [comment]
@johnknoop #2328

**Comment by johnknoop** [comment]
@Joe4evr Absolutely, I've already suppressed the warnings, so this is nothing that really blocks us. I just wanted to let you guys know so you can make the new nullable feature even more awesome.

Also, I might be a bit spoiled when it comes to flow analysis since I work a lot in TypeScript and it's just great when it comes to inference.

**Comment by Joe4evr** [comment]
TypeScript's flow analysis has had a lot more time to mature, though. But I certainly expect the C# LDT to keep improving the analysis rules in the future, so that devs can convey even more intent while also letting the compiler help more on the consuming side.

**Comment by Korporal** [comment]
> I've just enabled the new nullable reference types feature.
> 
> In this example, `myObj` is nullable. But shouldn't the compiler understand that myObj won't be null in this case, since I'm testing for it in the if clause?
> 
> ![image](https://user-images.githubusercontent.com/3394270/66725159-cc007880-ee2e-11e9-8720-04b72fd709b0.png)
> 
> I'm using .NET Core 3.0 in VS 2019.

How is the compiler to know that the value of `myObj` cannot be asynchronously changed on some other thread? for example you might have a time setup that manipulates `myObj`.

**Comment by Joe4evr** [comment]
@Korporal The compiler doesn't make any such assumptions.

**Comment by 333fred** [comment]
> https://github.com/dotnet/csharplang/issues/1581#issuecomment-544294729
The nullable feature assumes that if you check it, it's not going to change on a different thread. We also make this assumption on other things, like class properties and fields.

This is a compiler bug, I'm going to move to dotnet/roslyn.
